In [2]:
import pandas as pd
import matplotlib.pyplot as plt   
%matplotlib inline
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
import numpy as np
import statsmodels.formula.api as smf
import random

In [3]:
df = pd.read_csv('salary.csv')

In [4]:
df = df.dropna(how='any',axis=0)

In [12]:
y = df.salary
X_train, X_test, y_train, y_test = train_test_split(df, y, test_size=0.3, random_state=9)

salary  exprior  yearsworked  yearsrank  market  degree  otherqual  \
25   33680.0        0            3          3    0.88       1          0   
230  62211.0        0           21         11    1.14       1          0   
62   38333.0        0            7          2    0.80       1          0   
6    61008.0        0            2          2    1.33       1          0   
54   36308.0        3            2          2    0.80       1          0   

     position  male  Field  yearsabs  
25          1     1      4         1  
230         3     1      2         0  
62          2     1      3         2  
6           1     1      2         0  
54          1     1      1         0

In [6]:
model = smf.ols(formula='salary ~ yearsworked', data=X_train).fit() 
model.params

Intercept      39378.307467
yearsworked      846.938724
dtype: float64

In [7]:
print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:                 salary   R-squared:                       0.419
Model:                            OLS   Adj. R-squared:                  0.417
Method:                 Least Squares   F-statistic:                     257.5
Date:                Mon, 01 Apr 2019   Prob (F-statistic):           5.23e-44
Time:                        13:31:35   Log-Likelihood:                -3799.4
No. Observations:                 359   AIC:                             7603.
Df Residuals:                     357   BIC:                             7611.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept    3.938e+04    871.667     45.176      

### Does the model significantly predicts the dependent variable? Report the amount of variance explained (R^2) and significance value (p) to support your answer.

The model significantly predicts the dependent variable. This is shown from variance (R^2) 0.419 (higher value generally indicates a better fit) with a p value of 0.000... (less than 0.05 is considered to be statistically significant)


### What percentage of the variance in employees’ salaries is accounted for by the number of years they have worked?
75.9%



### What does the unstandardized coefficient (B or 'coef' in statsmodels) tell you about the relationship between Years Worked and Salary?
The coeffecient is 2840.5341 this means that when salary increases by 1, the number of salary increases by 2840.5341

### What do the 95% confidence intervals [0.025, 0.975] mean?

It shows the amount you would expect at a 95% confidence level. 

### Calculate the expected salary for someone with 12 years’ work experience.

In [8]:
predictions = model.predict(X_train)
print('Expected salary is ' + str(predictions[12]))

Expected salary is 41919.123639120146


### Calculate the expected salary for someone with 80 years’ work experience. Are there any problems with this prediction? If so, what are they?

In [17]:
pred = pd.DataFrame({'yearsworked': [80]})
model.predict(pred)


0    107133.405395
dtype: float64

### We have only looked at the number of years an employee has worked. What other employee characteristics might influence their salary?
yearsworked<br>
yearsrank<br>
position

In [21]:
df.corr()

salary   exprior  yearsworked  yearsrank    market    degree  \
salary       1.000000  0.119989     0.623589   0.610706  0.407214  0.044554   
exprior      0.119989  1.000000    -0.245788  -0.065893 -0.035202 -0.049165   
yearsworked  0.623589 -0.245788     1.000000   0.813456 -0.070740  0.028122   
yearsrank    0.610706 -0.065893     0.813456   1.000000 -0.026869  0.006347   
market       0.407214 -0.035202    -0.070740  -0.026869  1.000000  0.036456   
degree       0.044554 -0.049165     0.028122   0.006347  0.036456  1.000000   
otherqual    0.167137  0.264655    -0.040157   0.007232  0.021641 -0.214654   
position     0.702122  0.118031     0.746311   0.484891 -0.013116  0.043335   
male         0.361564  0.047276     0.278286   0.237397  0.181401  0.061419   
Field       -0.499316 -0.046854    -0.264208  -0.265620 -0.223887 -0.099077   
yearsabs    -0.069630  0.132022     0.056113   0.035962 -0.167177  0.029447   

             otherqual  position      male     Field  yearsabs  
salary        0.167137  0.702122  0.361564 -0.499316 -0.069630  
exprior       0.264655  0.118031  0.047276 -0.046854  0.132022  
yearsworked  -0.040157  0.746311  0.278286 -0.264208  0.056113  
yearsrank     0.007232  0.484891  0.237397 -0.265620  0.035962  
market        0.021641 -0.013116  0.181401 -0.223887 -0.167177  
degree       -0.214654  0.043335  0.061419 -0.099077  0.029447  
otherqual     1.000000  0.160941  0.016082 -0.076202 -0.017792  
position      0.160941  1.000000  0.317390 -0.292313  0.030521  
male          0.016082  0.317390  1.000000 -0.130599 -0.622058  
Field        -0.076202 -0.292313 -0.130599  1.000000  0.000778  
yearsabs     -0.017792  0.030521 -0.622058  0.000778  1.000000

### How does your model compare when running it on the test set - what is the difference in the Root Mean Square Error (RMSE) between the training and test sets?

In [ ]:
X_test = test['yearsworked']
y_test= test['salary']


In [ ]:
test_model = sm.OLS(y_test,X_test).fit()

In [ ]:
print_test_model = test_model.summary()
print(print_test_model)